src/vma/dev/cq_mgr.cpp

```diff
@@ -1600,3 +1599,356 @@ void cq_mgr::mlx5_init_cq()
 	m_mlx5_cqes = (volatile struct mlx5_cqe64 (*)[])(uintptr_t)m_mlx5_cq->active_buf->buf;
 }
 #endif // DEFINED_VMAPOLL
+
+#ifdef HAVE_INFINIBAND_MLX5_HW_H
+cq_mgr_mlx5::cq_mgr_mlx5(ring_simple* p_ring, ib_ctx_handler* p_ib_ctx_handler, uint32_t cq_size, struct ibv_comp_channel* p_comp_event_channel, bool is_rx):
+	cq_mgr(p_ring, p_ib_ctx_handler, cq_size, p_comp_event_channel, is_rx)
+	, m_cq_size(cq_size)
+	, m_cq_cons_index(0)
+	, m_cqes(NULL)
+	, m_cq_dbell(NULL)
+	, m_rx_hot_buffer(NULL)
+	,m_rq(NULL)
+	,m_p_rq_wqe_idx_to_wrid(NULL)
+{
+	cq_logfunc("");
+
+	struct ibv_cq *ibcq = m_p_ibv_cq;
+	struct mlx5_cq* mlx5_cq = _to_mxxx(cq, cq);
+	m_cq_dbell = mlx5_cq->dbrec;
+	m_cqes = (volatile struct mlx5_cqe64 (*)[])(uintptr_t)mlx5_cq->active_buf->buf;
+}
+
+cq_mgr_mlx5::~cq_mgr_mlx5()
+{
+	cq_logfunc("");
+}
+
+inline volatile struct mlx5_cqe64* cq_mgr_mlx5::check_cqe(void)
+{
+
+	volatile struct mlx5_cqe64 *cqes = *m_cqes;
+	volatile struct mlx5_cqe64 *cqe= &cqes[m_cq_cons_index & (m_cq_size - 1)];
+	uint16_t idx = m_cq_cons_index & m_cq_size;
+	uint8_t op_own = cqe->op_own;
+	uint8_t op_owner = ((op_own) & MLX5_CQE_OWNER_MASK);
+	uint8_t op_code = (((op_own) & 0xf0) >> 4);
+
+	if ((op_owner != (!!(idx))) || (op_code == MLX5_CQE_INVALID)) {
+		return NULL;/* No CQE. */
+	}
+
+	return cqe;
+}
+
+volatile struct mlx5_cqe64*	cq_mgr_mlx5::check_error_completion(uint8_t op_own)
+{
+	switch (op_own >> 4) {
```

Comment:
There are many places where you still have something like `(op_own >> 4)`.
It would be much more readable if you had something like `(MLX5_CQE_OPCODE(op_own))` where:

`#define MLX5_CQE_OPCODE(owner) ((owner) >> 4)`